nf_core/pipelines/download.py

```diff
@@ -1340,16 +1369,15 @@ def singularity_download_image(
                         fh.write(data)
 
             # Rename partial filename to final filename
-            os.rename(output_path_tmp, output_path)
+            os.rename(output_path_tmp, out_path)
 
-            # Copy cached download if we are using the cache
+            # Copy download to cache if one is defined
             if cache_path:
-                log.debug(f"Copying {container} from cache: '{os.path.basename(out_path)}'")
-                progress.update(task, description="Copying from cache to target directory")
-                shutil.copyfile(cache_path, out_path)
+                progress.update(task, description="Copying from target directory to cache")
+                self.singularity_copy_image(container, out_path, cache_path)
```

Comment:
You need to check now, if the cache_path actually exists on the file system:
`os.path.exists(cache_path)`.

Previously, the `get_singularity_images()` function ensured, that an image was actually present in cache before adding it to the list of images that are already cached.

```Python
  if cache_path and os.path.exists(cache_path):
      containers_cache.append((container, out_path, cache_path))
```
Hence, it was safe to copy it without further checks. Now, the `cache_path` is just created from the environment variable without further checks if the defined directory actually exists in the file system and is writable. Both needs to be done either here just for the cache path or preferably inside the generic copy function - better safe than sorry.